tsdb/engine/bz1/bz1.go

```diff
@@ -0,0 +1,597 @@
+		}
+		if err := e.writeFields(tx, measurementFieldsToSave); err != nil {
+			return fmt.Errorf("write fields: %s", err)
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// Write points to the WAL.
+	if err := e.PointsWriter.WritePoints(points); err != nil {
+		return fmt.Errorf("write points: %s", err)
+	}
+
+	return nil
+}
+
+// writeSeries writes a list of series to the metadata.
+func (e *Engine) writeSeries(tx *bolt.Tx, a []*tsdb.SeriesCreate) error {
+	// Ignore if there are no series.
+	if len(a) == 0 {
+		return nil
+	}
+
+	// Marshal and insert each series into the metadata.
+	b := tx.Bucket([]byte("series"))
+	for _, sc := range a {
+		// Marshal series into bytes.
+		data, err := sc.Series.MarshalBinary()
+		if err != nil {
+			return fmt.Errorf("marshal series: %s", err)
+		}
+
+		// Insert marshaled data into appropriate key.
+		if err := b.Put([]byte(sc.Series.Key), data); err != nil {
+			return fmt.Errorf("put: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// writeFields writes a list of measurement fields to the metadata.
+func (e *Engine) writeFields(tx *bolt.Tx, m map[string]*tsdb.MeasurementFields) error {
+	// Ignore if there are no fields to save.
+	if len(m) == 0 {
+		return nil
+	}
+
+	// Persist each measurement field in the map.
+	b := tx.Bucket([]byte("fields"))
+	for k, f := range m {
+		// Marshal field into bytes.
+		data, err := f.MarshalBinary()
+		if err != nil {
+			return fmt.Errorf("marshal measurement field: %s", err)
+		}
+
+		// Insert marshaled data into key.
+		if err := b.Put([]byte(k), data); err != nil {
+			return fmt.Errorf("put: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// WriteIndex writes marshaled points to the engine's underlying index.
+func (e *Engine) WriteIndex(pointsByKey map[string][][]byte) error {
+	return e.db.Update(func(tx *bolt.Tx) error {
+		for key, values := range pointsByKey {
+			if err := e.writeIndex(tx, key, values); err != nil {
+				return fmt.Errorf("write: key=%x, err=%s", key, err)
+			}
+		}
+		return nil
+	})
+}
+
+// writeIndex writes a set of points for a single key.
+func (e *Engine) writeIndex(tx *bolt.Tx, key string, a [][]byte) error {
+	// Ignore if there are no points.
+	if len(a) == 0 {
+		return nil
+	}
+
+	// Create or retrieve series bucket.
+	bkt, err := tx.Bucket([]byte("points")).CreateBucketIfNotExists([]byte(key))
+	if err != nil {
+		return fmt.Errorf("create series bucket: %s", err)
+	}
+	c := bkt.Cursor()
+
+	// Determine time range of new data.
+	tmin, tmax := int64(btou64(a[0][0:8])), int64(btou64(a[len(a)-1][0:8]))
+
+	// If tmin is after the last block then append new blocks.
+	//
+	// This is the optimized fast path. Otherwise we need to merge the points
+	// with existing blocks on disk and rewrite all the blocks for that range.
+	if k, v := c.Last(); k == nil || int64(btou64(v[0:8])) < tmin {
+		if err := e.writeBlocks(bkt, a); err != nil {
+			return fmt.Errorf("append blocks: %s", err)
+		}
+	}
```

Comment:
This is the fast path. It grabs the last block in the bucket, reads the first 8 bytes (which is the ending timestamp), and if the minimum time for the inserted points is greater than that then it can simply append new blocks.

Otherwise it's a big PITA to read out and reinsert blocks as you can see below.